File: javascript/packages/ethereum/src/FakeAttestationLookup.js
/**
 * In-memory attestation lookup for tests. No network. Implementations register
 * attestations by UID and optionally by wallet (recipient) for getDelegationsForWallet.
 */

const ZERO_UID = '0x0000000000000000000000000000000000000000000000000000000000000000';

function toRecord(att) {
  return {
    id: att.id ?? att.uid,
    attester: att.attester ?? '',
    recipient: att.recipient ?? '',
    schema: att.schema ?? att.schemaId ?? '',
    refUID: att.refUID ?? ZERO_UID,
    data: att.data ?? '0x',
    revoked: att.revoked === true,
    expirationTime: typeof att.expirationTime === 'number' ? att.expirationTime : parseInt(att.expirationTime || '0', 10) || 0,
    revocationTime: typeof att.revocationTime === 'number' ? att.revocationTime : parseInt(att.revocationTime || '0', 10) || 0
  };
}

/**
 * Create a fake lookup. Add attestations with addAttestation(record).
 * Optionally set leaves per wallet with setDelegationsForWallet(networkId, walletAddress, attestationRecords).
 * @returns Fake lookup + { addAttestation, setDelegationsForWallet } for test setup
 */
export function createFakeAttestationLookup() {
  const byUid = new Map();
  const byWallet = new Map();
  const networkIds = new Set();

  function key(networkId, wallet) {
    return `${(networkId || '').toLowerCase()}:${(wallet || '').toLowerCase()}`;
  }

  return {
    addAttestation(record, networkId) {
      const r = toRecord(record);
      const uid = (r.id || '').toLowerCase();
      if (uid) byUid.set(uid, r);
      if (networkId) networkIds.add((networkId || '').toLowerCase());
      return this;
    },

    getSupportedNetworks() {
      return Array.from(networkIds);
    },

    setDelegationsForWallet(networkId, walletAddress, attestationRecords) {
      if (networkId) networkIds.add((networkId || '').toLowerCase());
      const k = key(networkId, walletAddress);
      byWallet.set(k, (attestationRecords || []).map(toRecord));
      for (const r of byWallet.get(k)) {
        const uid = (r.id || '').toLowerCase();
        if (uid) byUid.set(uid, r);
      }
      return this;
    },

    async getDelegationsForWallet(networkId, walletAddress) {
      const k = key(networkId, walletAddress);
      return byWallet.get(k) ?? [];
    },

    async getAttestation(networkId, uid) {
      const id = (uid || '').toLowerCase();
      return byUid.get(id) ?? null;
    }
  };
}

File: dotnet/src/Zipwire.ProofPack.Ethereum/ProofPack/FakeAttestationLookup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Zipwire.ProofPack.Ethereum;

/// <summary>
/// In-memory attestation lookup for tests. Register attestations by UID and by wallet (recipient)
/// for GetDelegationsForWallet. No network calls.
/// </summary>
public sealed class FakeAttestationLookup : IAttestationLookup
{
    private readonly Dictionary<string, AttestationRecord> _byUid = new(StringComparer.OrdinalIgnoreCase);
    private readonly Dictionary<string, List<AttestationRecord>> _byWallet = new(StringComparer.OrdinalIgnoreCase);
    private readonly HashSet<string> _networkIds = new(StringComparer.OrdinalIgnoreCase);

    private static string Key(string? networkId, string? wallet)
    {
        return $"{(networkId ?? string.Empty).ToLowerInvariant()}:{(wallet ?? string.Empty).ToLowerInvariant()}";
    }

    /// <summary>
    /// Adds an attestation record so GetAttestationAsync can return it by UID.
    /// </summary>
    /// <param name="record">The attestation record (Id must be set).</param>
    /// <param name="networkId">Optional; when set, adds to supported networks.</param>
    public FakeAttestationLookup AddAttestation(AttestationRecord record, string? networkId = null)
    {
        if (record?.Id != null)
        {
            _byUid[record.Id.ToLowerInvariant()] = record;
        }

        if (!string.IsNullOrEmpty(networkId))
        {
            _networkIds.Add(networkId.ToLowerInvariant());
        }

        return this;
    }

    /// <summary>
    /// Sets the list of IsDelegate attestation records returned for a wallet on a network.
    /// Also registers each record for GetAttestationAsync by UID.
    /// </summary>
    public FakeAttestationLookup SetDelegationsForWallet(string networkId, string walletAddress, IReadOnlyList<AttestationRecord> records)
    {
        if (!string.IsNullOrEmpty(networkId))
        {
            _networkIds.Add(networkId.ToLowerInvariant());
        }

        var list = (records ?? Array.Empty<AttestationRecord>()).ToList();
        _byWallet[Key(networkId, walletAddress)] = list;
        foreach (var r in list)
        {
            if (!string.IsNullOrEmpty(r?.Id))
            {
                _byUid[r.Id.ToLowerInvariant()] = r;
            }
        }

        return this;
    }

    /// <inheritdoc />
    public IReadOnlyList<string>? GetSupportedNetworks()
    {
        return _networkIds.ToList();
    }

    /// <inheritdoc />
    public Task<IReadOnlyList<AttestationRecord>> GetDelegationsForWalletAsync(
        string networkId,
        string walletAddress,
        CancellationToken cancellationToken = default)
    {
        var key = Key(networkId, walletAddress);
        var list = _byWallet.TryGetValue(key, out var records) ? records : (IReadOnlyList<AttestationRecord>)Array.Empty<AttestationRecord>();
        return Task.FromResult(list);
    }

    /// <inheritdoc />
    public Task<IReadOnlyList<AttestationRecord>> GetAttestationsForWalletBySchemasAsync(
        string networkId,
        string walletAddress,
        IReadOnlyList<string> schemaIds,
        CancellationToken cancellationToken = default)
    {
        var key = Key(networkId, walletAddress);
        if (!_byWallet.TryGetValue(key, out var records) || schemaIds == null || schemaIds.Count == 0)
        {
            return Task.FromResult<IReadOnlyList<AttestationRecord>>(Array.Empty<AttestationRecord>());
        }

        var set = new HashSet<string>(schemaIds.Where(s => !string.IsNullOrEmpty(s)), StringComparer.OrdinalIgnoreCase);
        var filtered = records.Where(r => set.Contains(r.Schema ?? string.Empty)).ToList();
        return Task.FromResult<IReadOnlyList<AttestationRecord>>(filtered);
    }

    /// <inheritdoc />
    public Task<AttestationRecord?> GetAttestationAsync(
        string networkId,
        string uid,
        CancellationToken cancellationToken = default)
    {
        var id = (uid ?? string.Empty).ToLowerInvariant();
        var found = _byUid.TryGetValue(id, out var record);
        return Task.FromResult(found ? record : (AttestationRecord?)null);
    }
}
